### ISSUE TITLE
JUnit Output

### ISSUE THREAD

**Issue by rodnaph** [jonase/eastwood#160]
I'm interested in running Eastwood inside a Jenkins build and having the results output for parsing by a post-build publisher (Junit seems to be the easiest to target).

At the moment the way support for parsing warning information is set up seems to be via using a text file with the :out parameter - is this how all planned integration should be achieved or would there be any interest in a patch for extensible output formatting?

**Comment by jafingerhut** [comment]
There are currently a few options for selecting the format of the output text.  See the description of the :warning-format key here: https://github.com/jonase/eastwood#editor-support

If you would prefer a different text output format, let me know what you are looking for.

Eastwood can also be invoked from within a REPL, and either print its output as text, or return it as a sequence of maps with keys like :line :column :file etc.  See here: https://github.com/jonase/eastwood#running-eastwood-in-a-repl

**Comment by rodnaph** [comment]
Hi Andy, yes I saw that option, but I'm looking to produce XML output from the result data and couldn't see how I would bend the line-by-line approach to create structured data.  I've managed to wrap it pretty easily...

https://github.com/Owsy/eastwood-junit

In the same way this project wraps test output I suppose...

https://github.com/arohner/lein-test-out

Anyway, if there is interest in a patch I can supply one, otherwise no worries.

**Comment by jafingerhut** [comment]
Would adding another value for :warning-format be sufficient, e.g. {:warning-format :junit} ?

**Comment by rodnaph** [comment]
I don't think that would be enough (based on my limited understanding of Eastwood anyway).  Correct me if I'm wrong but the current output is written warning by warning eagerly as the codebase is linted, is this correct?

For example, JUnit output is structured as follow:

```xml
<?xml version="1.0"?>

**Comment by rodnaph** [comment]
I don't think so, based on my limited understanding of Eastwood anyway.  Correct me if I'm wrong but it seems that currently the formatted output is written warning by warning eagerly as the codbase is linted, is this correct?

So JUnit for example needs to be an XML document like...

```xml
<?xml version="1.0"?>
<testsuites>
  <testsuite package="eastwood.lint">
    <testcase name="eastwood.lint.line3">
      <failure type="linter" message="the failure message" />
    </testcase>
  </testsuite>
</textsuites>
```

So I can't just write out the warnings as they occur because I also need to include the XML header, group by package/namespace etc...  Now, I could probably fudge something if I could get a callback in there but it'd be far far from ideal.

Is my understanding correct or is the output formatting more flexible then I thought?  My suggested implementation for formatting structured data would probably need to a hook after all warnings had been created.

**Comment by jafingerhut** [comment]
Ah, I see now.  Yes, you are correct that the existing code for the :warning-format option makes it straightforward to change the format of individual warnings, one by one, but does not make it easy to change the format of the output as a whole.  I will think about this a bit more, later, and add a comment here if I think of anything.
